src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.Analyzer.cs

```diff
@@ -0,0 +1,365 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.CodeRefactorings.InvertIf
+{
+    internal sealed partial class CSharpInvertIfCodeRefactoringProvider
+    {
+        private sealed class Analyzer : Analyzer<IfStatementSyntax>
+        {
+            protected override string GetTitle()
+            {
+                return CSharpFeaturesResources.Invert_if;
+            }
+
+            protected override SyntaxNode GetRootWithInvertIfStatement(
+                Document document,
+                SemanticModel semanticModel,
+                IfStatementSyntax ifStatement,
+                InvertIfStyle invertIfStyle,
+                SyntaxNode subsequenceSingleExitPointOpt,
+                CancellationToken cancellationToken)
+            {
+                var generator = SyntaxGenerator.GetGenerator(document);
+                var syntaxFacts = CSharpSyntaxFactsService.Instance;
+
+                var ifNode = ifStatement;
+
+                var negatedCondition = (ExpressionSyntax)Negator.Negate(ifStatement.Condition, generator, syntaxFacts, semanticModel, cancellationToken);
+                var root = semanticModel.SyntaxTree.GetRoot(cancellationToken);
+
+                switch (invertIfStyle)
+                {
+                    case InvertIfStyle.Normal:
+                        {
+                            // For single line statement, we swap the TrailingTrivia to preserve the single line
+                            StatementSyntax newIfNodeStatement = null;
+                            ElseClauseSyntax newElseStatement = null;
+                            IfStatementSyntax oldIfStatement = ifStatement;
+
+                            var hasNewLineAfterClosingBrace = oldIfStatement.Statement.GetTrailingTrivia().Any(trivia => trivia.Kind() == SyntaxKind.EndOfLineTrivia);
```

Comment:
this name is strange.  It implies there is a closing brace... but how do you know that's the case?

Reply:
These are all existing code untouched. @chborl can weigh in to clarify the intention. I'll try to review those parts once we're done with elseless.. I want to simplify VB part to not need semantic update, (TestTryToParenthesizeAwkwardSyntaxInsideSingleLineLambdaMethod) e.g. a simple syntax exact match but I'm not sure if this is the only case.

Reply:
> I want to simplify VB part to not need semantic update

I'm ok with that.  I don't think all the code complexity carries its weight in terms of the value it provides the end user.

Reply:
(resolved)